### DISCUSSION TITLE
does ANYONE knows how to add custom notes (danger and warning notes like in Tricky) in KE 1.8?

### DISCUSSION THREAD

**Discussion by AlexWindy1** [KadeArchive/Kade-Engine#2916] [Q&A]
Hello codder, I have a big problem with my KE 1.8 mod I'm making, I want to make a song have gimmicks, Danger notes to be more precise, but idk how on 1.8, all the tutorials that exists are on older versions of KE, and it has changed since then, please tell me someone knows how to add those custom notes in KE 1.8, I really need, I have made so much on 1.8 yet, I can't start all over on an older version...

thanks in advance to anyone who might know the answer

**Comment by BoloVEVO** [comment]
Kade Engine is fkin dead. KadeDev doesn't mantain this anymore but I love this engine above psych and I'm too generous.
I managed to add custom notes to my custom KE 1.8 fork. I surely recommed you use my Kade Engine's fork that fixes many bugs this Kade Engine version have.
1. Go to Note.hx and define your note type var and include it in Note function new.
![image](https://user-images.githubusercontent.com/88955176/164402556-96c6b1dd-4d81-45d1-ada6-c3ed835014db.png)
![image](https://user-images.githubusercontent.com/88955176/164406708-4d71c597-eb8c-4491-952d-6cfb76646044.png)

2. Now go to NoteSkinHelper.hx and make a function like this I left you the screenshit as a template for your custom note. 
**(You can skip this step if your note asset is unique regardless the noteskin the user have).**
Note: No need to say it but you need to create a folder called 'custom_note_skins' inside shared folder (you can change it if you want ,just did it to be organized and not fill shared folder with bullshit).
![image](https://user-images.githubusercontent.com/88955176/164404136-c6040c18-3be5-49ae-9321-5ae68a424247.png)
![image](https://user-images.githubusercontent.com/88955176/164404044-8c2ab129-14e1-4a39-91b8-a1db0d346651.png)
![image](https://user-images.githubusercontent.com/88955176/164404103-0dab9be7-4c03-4364-9526-3cbe826dbaa5.png)
4. Return to Note.hx and go around the note assets part. Depending on what stage type you want (pixel or normal) make a switch case.
'default' being the normal notes and 'electro' your custom notes (obviously you can change the name to whathever you want)
![image](https://user-images.githubusercontent.com/88955176/164405596-450e69c2-ebc2-4f0a-aaf9-1bc50fd3f210.png)

Now the custom Note asset part is over.

Now we're goin to PlayState.hx 


4.5. If you did 2nd step. Define your custom note skin and give it the noteskin value in create function
![image](https://user-images.githubusercontent.com/88955176/164485618-c651ecc4-1af6-45fa-82a3-e60ebe8907ae.png)
![image](https://user-images.githubusercontent.com/88955176/164486027-972af58c-1716-4c63-9f37-f6f3cabab316.png)

5. Assign notekind var to a songNotes. Preferible to songNotes[5] to follow an order.
![image](https://user-images.githubusercontent.com/88955176/164406953-1ab6d248-db95-4ff5-925f-84219db7383c.png)
![image](https://user-images.githubusercontent.com/88955176/164407099-dbb40b4a-2491-4d04-93a1-0428093b7d0f.png)
6. In same PlayState.hx go to note mechanics part around here and make a switch function.
Do you see a code below else if condition in your PlayState.hx? Mark that as 'default:' to your switch function. No need to create an entire code with your custom case because it has same miss function as normal with difference it will instakill u.
![image](https://user-images.githubusercontent.com/88955176/164408303-da4b9969-4b9e-4392-9047-5a5d6d772a63.png)
```
else if ((daNote.mustPress && !PlayStateChangeables.useDownscroll || daNote.mustPress && PlayStateChangeables.useDownscroll)
					&& daNote.mustPress
					&& Conductor.songPosition > noteKillOffset + daNote.strumTime
					&& songStarted)
{
       switch (daNote.noteKind)
      {
            default: // The note code you see below the else if condition here!
      }
 }
```
7. Now in PlayState.hx search noteMiss function and write your danger note condition. So basically when you miss your health is set to 0 and when your health is set to 0 u die bruh.
![image](https://user-images.githubusercontent.com/88955176/164412128-cf7f3766-5401-491c-bb43-af2d8403d7c5.png)

**Custom Note type mechanic is over**

Next steps will be inside ChartingState.hx
8. Define this two vars.
![image](https://user-images.githubusercontent.com/88955176/164412946-14f05bf5-5826-439a-a539-46dbd0274ec7.png)
9. If you didn't stepped Step 2 put this in create function to update note asset regarding user noteskin.
![image](https://user-images.githubusercontent.com/88955176/164414451-374a34c9-a1d6-4d32-9fcf-d2f1792a0335.png)

10 Go around this condition (You can search it in visual code search section) and make this condition. So to change your note type just press Z and X and you will place the custom note you want.
![image](https://user-images.githubusercontent.com/88955176/164413262-4f8e1e7b-2aab-4886-9beb-5ab17fd762e6.png)
11. Go to addNote function and insert the notekind var. This must looks like this.
![image](https://user-images.githubusercontent.com/88955176/164413573-e83f5ea8-e4ab-4bbd-bac0-9cff71a7e4a2.png)
![image](https://user-images.githubusercontent.com/88955176/164413702-5ddac652-f745-45b3-8801-ead22b9c4000.png)
![image](https://user-images.githubusercontent.com/88955176/164413736-7c4e367e-4be2-4255-8c32-1d2a3280447d.png)

Now we are over. Just open the game go to Charting menu and press Z or X to switch between normal notes and danger notes and you'll be putting danger notes in your chart.

If your game crashes when you add a note is because your note asset is bad or you made a mistake assigning note assets in Note.hx during Step 3.

If same occurs when you press the custom note is related to step 6.

Don't forget to debug your game in Visual Code Studio to find the source of the problem.

Enjoy my boi!

**Reply by AlexWindy1** [reply]
I did everything you said, but it doesn't launch, it gives me that message error in the CMD, do you know how it happened and how to fix it?
![image](https://user-images.githubusercontent.com/104106683/164470473-036e0149-385f-452b-8d96-4a4fd4904709.png)

**Reply by BoloVEVO** [reply]
> I did everything you said, but it doesn't launch, it gives me that message error in the CMD, do you know how it happened and how to fix it? ![image](https://user-images.githubusercontent.com/104106683/164470473-036e0149-385f-452b-8d96-4a4fd4904709.png)

Send screenshot of that line to see what happened pls.

**Reply by AlexWindy1** [reply]
there we go
![image](https://user-images.githubusercontent.com/104106683/164479409-6b929de7-fafa-4cad-b936-2fcef38f5f2c.png)

**Reply by AlexWindy1** [reply]
and the weird thing is that, if I remove "noteKind" the game lauches, so... idk

**Reply by BoloVEVO** [reply]
You replaced directly notekind in 'isAlt' bool place. Type isAlt before noteKind like this:
![image](https://user-images.githubusercontent.com/88955176/164480912-e632adc2-ca53-4421-9bd5-0a396868636c.png)

**Reply by AlexWindy1** [reply]
oh, okay lemme try

**Reply by AlexWindy1** [reply]
but, on the screenshot you sent me, there is no isAlt function?

**Reply by AlexWindy1** [reply]
Nope, still have the same problem
![image](https://user-images.githubusercontent.com/104106683/164481703-9d25be8c-32b6-41c4-8a3d-465e5700254a.png)

**Reply by AlexWindy1** [reply]
though, I look again through what I did, and I saw one function I didn't putted a capital "K" and when I did it, it worked, the game launched!

but now I have another problem, my custom arrows don't show up in the chart editor, and all I have is full grey Arrows, and I can tap on Z or X as much as I want, still have just grey arrows, though, in game, they all appear in colors

**Reply by AlexWindy1** [reply]
nevermind, they show up XD but idk if they work...

**Reply by AlexWindy1** [reply]
okay no, they don't work, they show in the chart editor, but in game, they are just normal arrows

**Reply by BoloVEVO** [reply]
> but, on the screenshot you sent me, there is no isAlt function?

Indeed, my bad. Forget about doing that. That's not the problem. I forgot one step inside ChartingState.hx . Inside updateGrid function insert noteKind field like this.

![image](https://user-images.githubusercontent.com/88955176/164483830-6bf2322e-9a9e-46cb-a52b-8fdfb675cab8.png)

**Reply by AlexWindy1** [reply]
mh, I still have the arrows in grey in the chart editor, and in-game the custom arrows still show up as Normal notes and act like normal notes too

**Reply by AlexWindy1** [reply]
also we are filling up that discussion... perhaps we should switch to somewhere else, you have discord? if yes, add me Windy#7057

**Reply by BoloVEVO** [reply]
> mh, I still have the arrows in grey in the chart editor, and in-game the custom arrows still show up as Normal notes and act like normal notes too

Did you press X and after that added a note?
Your custom note asset is unique or is regarding the user noteskin?
About the mechanic I gave you an example of a warning note (you miss one, you die)

**Reply by AlexWindy1** [reply]
yes, when I press X the custom arrow show up in the chart editor, but, when in-game, playing the song, they just show up in Normal notes,

and yes I think it is considered unique, I edited the notes in the flash file of them, then export it as a whole new ref sheet of them, and put them in a different folder etc...

and I did the mecanics, of the note that kills you if you don't touch it (even though it's when the player touch it that I want him to die)

but still they act like totally normal notes, if you miss them you lose health etc... also, when putting them in the chart editor, and then quit it then go back in, they show up in the chart editor as grey arrows, like all the others that are normal

**Reply by AlexWindy1** [reply]
I finally managed to get it to work, YES, just the mecanics now I have problem with

**Reply by BoloVEVO** [reply]
Apologies for the inactivity I completely forgot about this discussion. Did you check if it's right the note if statement on goodnoteHit (For tricky notes) or noteMiss (For Danger Notes)?

**Reply by AlexWindy1** [reply]
It's okay I'm glad to see you back ^^

No I didn't, what do I need to write for it? And where? Because it's not send on your screenshots at the beginning

**Reply by AlexWindy1** [reply]
okay, I managed to make warning notes (you miss, you die)
but for the instaKill notes, how do I change if it's pressed? because apparently the only place in the PlayState.hx where there is "if you tap the note" it's with the rating, but rating is useless for instakill notes

**Reply by AlexWindy1** [reply]
nevermind, I didn't found out, how to make neither the warning notes, neither the insta kill notes

I can do it by changing the health losing or giving of the normal notes, but can't with my custom one, I thought that by making the same coding as the normal notes but for the " case 'danger' " but no, don't work, I think my custom notes just show up in charting and in-game, but are just like, skins, not actual custom notes, urgh, I should have never started making my mod on KE 1.8 now I'm stuck for the gimmicks of the last song....

**Reply by BoloVEVO** [reply]
For instakill notes I recommed you using goodNoteHIt function, inside of this function inside the if statement (!note.wasgoodhit) make your if statement like this. Don't use rating switch (it crash ur game if you get a good). 
![image](https://user-images.githubusercontent.com/88955176/167058654-cdb0f522-512a-4748-b832-a931d83ff6fb.png)

**Reply by AlexWindy1** [reply]
Nope.... didn't worked, you sure I don't need a variable in the PlayState.hx for "danger" (or in your case the "instakill") for it to work? because, they keep acting as normal notes, I think we need a variable for the skin, like, one normal, and one "instakill"

**Reply by BoloVEVO** [reply]
It's very weird because I made my own if statement to my noteKind and the mechanic actually works. My custom notes don't give me a miss and also the purple bar goes to 0 when I hit it (also in the same goodnotehit function)

https://user-images.githubusercontent.com/88955176/167143144-739fe364-87da-445b-bc1f-e8be29b123de.mp4

**Reply by AlexWindy1** [reply]
maybe it is because I didn't do step 2? but I'm not sure, because anything else I done

**Reply by AlexWindy1** [reply]
okay, I got it to work, but now, all notes are my custom notes and when I try to put a normal note in chart editor; the game crashes

**Reply by AlexWindy1** [reply]
okay I succeeded at it, but uh, now, all my notes have the same function, of the last screen you sent me to give mecanics, for the " instakill " and I knowit's that, cause, well, I changed it to give health instead of taking some, and it does it, so idk, did I do something wrong again?

**Reply by AlexWindy1** [reply]
also, when I die, the song restart in the background, and continue, despite me being dead

**Reply by BoloVEVO** [reply]
Send some screenshots to see what did u do related to noteKinds in PlayState.hx pls.

**Reply by AlexWindy1** [reply]
Oki gimme a few minutes to screen all that

**Reply by AlexWindy1** [reply]
![image](https://user-images.githubusercontent.com/104106683/167274519-2e2c37d2-e1f9-4774-be2c-101036217d84.png)
![image](https://user-images.githubusercontent.com/104106683/167274525-859e0a99-7a0a-4cd6-b6ce-2736778dcbfb.png)
![image](https://user-images.githubusercontent.com/104106683/167274543-e1316a7d-34a9-47ac-acb2-67c4cf1deaea.png)
![image](https://user-images.githubusercontent.com/104106683/167274551-bf5685d3-6f63-462a-a06e-7bdcfdd671a6.png)

**Reply by AlexWindy1** [reply]
I have 5 of them only

**Reply by BoloVEVO** [reply]
Did you do step 6 correctly? Pls send me a screenshot of that part.

**Reply by AlexWindy1** [reply]
I did, but when doing it it broke the game and no notes were lowering my health when I miss them, whetever they were my custom or normal notes

**Reply by AlexWindy1** [reply]
but I'll try again okay, but I have doubts

**Reply by AlexWindy1** [reply]
![image](https://user-images.githubusercontent.com/104106683/167275352-059e305d-f681-4ed0-9b1f-2fd8c5aad7c9.png)
on that second screenshot it's where I closed the bracket of the switch function
![image](https://user-images.githubusercontent.com/104106683/167275390-134e7222-3f9a-4b9f-9be1-2dd1a966b1b0.png)

and it's still broken, with it added, I don't lose health when missing a note, and all the notes, normal or danger insta kill me

**Reply by BoloVEVO** [reply]
![image](https://user-images.githubusercontent.com/88955176/167277000-5b299e91-fce2-4749-851f-0061e89676c1.png)
Why the ` default:`  code is not indented to the right?

**Reply by AlexWindy1** [reply]
because, if it does, it's then in the "switch" line between the bracket, wouln't that cause a problem? plus i don't want to move each line under the default, would take me quite some time to do, there is a lot

**Reply by BoloVEVO** [reply]
It must be indented to the right to work otherwise it won't. And it's easy to indent it to the right just select all the code you want to indent to the right and press TAB.

**Reply by AlexWindy1** [reply]
oh I'll try
oki brb

**Reply by AlexWindy1** [reply]
*sigh* nope, still the same, don't lose health, when I die the song keep going, all notes insta kill me

**Reply by AlexWindy1** [reply]
maybe I made a mistake in Note.hx, or charting stat, or maybe I forgot something in the playstate

**Reply by AlexWindy1** [reply]
https://user-images.githubusercontent.com/104106683/167277508-af55543b-2c3d-4561-a4d8-e5658ac9df6c.mp4

**Reply by BoloVEVO** [reply]
Ok actually I have no clue. Could you send me your PlayState.hx file to try to fix it please.

**Reply by AlexWindy1** [reply]
sure, I can try

**Reply by AlexWindy1** [reply]
[code my mod.zip](https://github.com/KadeDev/Kade-Engine/files/8646162/code.my.mod.zip)

there we go

**Reply by AlexWindy1** [reply]
you saw the video?

**Reply by BoloVEVO** [reply]
Yea, I'll try to fix it later gotta go.

**Reply by AlexWindy1** [reply]
Oki doki see you later

**Reply by AlexWindy1** [reply]
hey Bolo, how's you?

**Reply by AlexWindy1** [reply]
Hello Bolo, hum... did you... found the source of the problem?

**Reply by AlexWindy1** [reply]
*sighs* I try re-doing everything again, from the start, but now my custom arrows don't appear, I can click on Z or X as much as I want, they don't show up, I put regular notes in the chart, and they are regular in-game, please, keep helping on all that, i'm losing my mind for real....
